blog/2023-10-22-constructs-api-design-2.md

```diff
@@ -0,0 +1,318 @@
+description: The second part in a series about designing safe APIs around mutation using the constructs framework
+authors: 
+  - rybickic
+tags: [winglang, cdk, constructs, api design, jsii, AWS CDK, mutability]
+hide_table_of_contents: true
+---
+
+Hey there! 👋
+My name's Chris and I'm a software engineer on the Wing Cloud team.
+Lately I've been helping out building Wing's compiler, and designing APIs for Wing's standard library.
+
+In the [first post](./2023-09-25-constructs-api-design.md) in this series, I introduced some of the challenges of designing APIs for [constructs](https://github.com/aws/constructs) and frameworks like AWS CDK, CDKTF, and cdk8s when mutation is involved.
+
+To recap, a construct can have public methods that mutate the objects' private state.
+But if this state is replaced or destroyed, then application code becomes more sensitive to the order of statements and method calls.
+This is usually undesirable when our constructs are modeling declarative information, like infrastructure configuration.
+
+To that end, we proposed two solutions for designing construct methods:
+
+1. Only add state, never subtract or update
+2. Document destructive APIs
+
+These measures are great for addressing a lot of our concerns with mutation.
+But as we'll see, mutation has wider effects than just the design of methods.
+
+## Sharing construct state through properties
+
+Another key capability of most constructs is that they can expose parts of their state for other constructs to use through **properties**.
+We can see how they're used in an example from the AWS CDK framework below, written in Wing:
+
+```js
+bring "aws-cdk-lib" as cdk;
+
+let table = new cdk.aws_dynamodb.Table(
+  partitionKey: {
+    name: "path",
+    type: cdk.aws_dynamodb.AttributeType.STRING
+  }
+) as "hits";
+
+let handler = new cdk.aws_lambda.Function(
+  runtime: cdk.aws_lambda.Runtime.NODEJS_14_X,
+  handler: "hitcounter.handler",
+  code: cdk.aws_lambda.Code.fromAsset("./lambda"),
+  environment: {
+    HITS_TABLE_NAME: table.tableName
+  }
+);
+```
+
+The construct named `Table` has a public property named `tableName` that stores the table's physical name for identifying it on AWS.
+The table's property `tableName` is passed as the `HITS_TABLE_NAME` environment variable so that the AWS Lambda function can use the table's dynamic name at runtime -- for example, to query the table (not shown).
+
+Any construct state that isn't meant to be a private implementation detail can be made public.
+But, as we've mentioned before, it's also possible for construct state to change after it was first initialized in the code.
+
+Uh oh - this smells like a recipe for problems.
+
+## When properties get stale
+
+Let's understand what causes properties to not play well with mutating methods through an example.
+I'll start by taking my `Flower` class from the previous post and adding options to specify the regions in the world where it's natively found:
+
+```js
```

Comment:
I'd add a note making it clear that this code is typescript, not Wing (since the previous sample is written in Wing)